### DISCUSSION TITLE
Sub device (BLE Mesh & ZigBee) works with homebridge-miot!

### DISCUSSION THREAD

**Discussion by 0x5e** [merdok/homebridge-miot#348] [Show and tell]
Inspired by https://sspai.com/post/68306 (In Chinese)
This article shows how to control ble mesh device via gateway using `miiocli` command line:
1. Get the gateway ip and token.
2. Get the ble mesh device id and model.
3. Using following command to set and get device properties via gateway:
```
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command set_properties "[{'did':'${DEVICE_ID}','value':True,'siid':2,'piid':1}]"
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command get_properties "[{'did':'${DEVICE_ID}','siid':2,'piid':1}]"
```

It sounds very easy, without ble hardware and driver in the computer.

So I have a try:
Add the device using gateway's ip, gateway's token, device's id, device's model. Save config and restart.
Although there's a warning says "the device model dosen't match", but it works.

My devices:
Gateway: 小爱音箱Pro, xiaomi.wifispeaker.lx06
Device: PTX蓝牙Mesh智能开关, 090615.switch.mesw2

Small issue: If I control the switch frequently, it won't respond and shows wrong status. I will test again later morning.

**Comment by merdok** [comment]
Oh, that is very interesting. If that is indeed so easy then controlling bluetooth devices over the gateway should be simple

**Comment by 0x5e** [comment]
Yeah, simple than I thought, and zigbee sub device works well too.
Seems that wifi device control command is compatible with sub device (zigbee & ble mesh) command.

Tested with:
Zigbee gateway: Aqara网关E1（青春版）, lumi.gateway.aqcn02
Device: Aqara 智能墙壁开关 E1, lumi.switch.b2lc04

And there's two issue now:
1. ble mesh device can't control frequently, otherwise it will fully not controlable, until I use Mi Home App to switch on and off. It's difficult to describe. And zigbee device didn't have this problem.
2. If I add two sub devices in one gateway, homebridge will raise the error says they have the same uuid. So currently only support one sub device in one gateway.

**Comment by 0x5e** [comment]
I tried to use `miiocli` to switch on and off, it works well. So I believe homebridge-miot plugin can solve as well. Maybe just cache problem. After national holiday back to home, I will continue to find out why.

```
GATEWAY_IP="192.168.x.x"
GATEWAY_TOKEN="xxxxxx"
DEVICE_ID="xxxxxxx"
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command set_properties "[{'did':'${DEVICE_ID}','value':True,'siid':2,'piid':1}]"
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command get_properties "[{'did':'${DEVICE_ID}','siid':2,'piid':1}]"
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command set_properties "[{'did':'${DEVICE_ID}','value':False,'siid':2,'piid':1}]"
miiocli device --ip $GATEWAY_IP --token $GATEWAY_TOKEN raw_command get_properties "[{'did':'${DEVICE_ID}','siid':2,'piid':1}]"

# Output
Running command raw_command
[{'did': 'xxxx', 'siid': 2, 'piid': 1, 'code': 1, 'ts': 1664773213, 'ret': 0}]
Running command raw_command
[{'did': 'xxxx', 'siid': 2, 'piid': 1, 'value': True, 'code': 0, 'ts': 1664773467, 'ret': 0}]
Running command raw_command
[{'did': 'xxxx', 'siid': 2, 'piid': 1, 'code': 1, 'ts': 1664773467, 'ret': 0}]
Running command raw_command
[{'did': 'xxxx', 'siid': 2, 'piid': 1, 'value': False, 'code': 0, 'ts': 1664773471, 'ret': 0}]
```

**Reply by merdok** [reply]
I think that should work out of the box with miot also, just try to use the miot CLI and let me know if it works, the command structure is very similar.

**Reply by 0x5e** [reply]
Sure, I'm not at home this week, I will feedback later

**Reply by merdok** [reply]
Thanks 👍

**Comment by 0x5e** [comment]
miot CLI works at most time, but sometime failed with `ERROR  user ack timeout`.
I execute 4 commands in one shot: turn off, get status, turn on, get status.
Compared with miot and miiocli, miiocli almost never failed, but miot almost failed at the third command every time.

```
miot send $GATEWAY_IP --token $GATEWAY_TOKEN set_properties "[{\"did\":\"$DEVICE_ID\",\"value\":false,\"siid\":2,\"piid\":1}]"
miot send $GATEWAY_IP --token $GATEWAY_TOKEN get_properties "[{\"did\":\"$DEVICE_ID\",\"siid\":2,\"piid\":1}]"
miot send $GATEWAY_IP --token $GATEWAY_TOKEN set_properties "[{\"did\":\"$DEVICE_ID\",\"value\":true,\"siid\":2,\"piid\":1}]"
miot send $GATEWAY_IP --token $GATEWAY_TOKEN get_properties "[{\"did\":\"$DEVICE_ID\",\"siid\":2,\"piid\":1}]"

# Output
 INFO  Connecting to device at  192.168.x.x
 INFO  Device found! Sending command: set_properties [{"did":"xxx","value":false,"siid":2,"piid":1}]
(Protocol) Call 192.168.x.x: set_properties - [{"did":"xxx","value":false,"siid":2,"piid":1}] - {"retries":1,"timeout":5000}
 SUCCESS  Response from device -> [{"did":"xxx","siid":2,"piid":1,"code":1,"ts":1665230823,"ret":0}]
 INFO  Connecting to device at  192.168.x.x
 INFO  Device found! Sending command: get_properties [{"did":"xxx","siid":2,"piid":1}]
(Protocol) Call 192.168.x.x: get_properties - [{"did":"xxx","siid":2,"piid":1}] - {"retries":1,"timeout":5000}
 SUCCESS  Response from device -> [{"did":"xxx","siid":2,"piid":1,"value":false,"code":0,"ts":1665230827,"ret":0}]
 INFO  Connecting to device at  192.168.x.x
 INFO  Device found! Sending command: set_properties [{"did":"xxx","value":true,"siid":2,"piid":1}]
(Protocol) Call 192.168.x.x: set_properties - [{"did":"xxx","value":true,"siid":2,"piid":1}] - {"retries":1,"timeout":5000}
 ERROR  user ack timeout
 INFO  Connecting to device at  192.168.x.x
 INFO  Device found! Sending command: get_properties [{"did":"xxx","siid":2,"piid":1}]
(Protocol) Call 192.168.x.x: get_properties - [{"did":"xxx","siid":2,"piid":1}] - {"retries":1,"timeout":5000}
 SUCCESS  Response from device -> [{"did":"xxx","siid":2,"piid":1,"value":false,"code":0,"ts":1665230831,"ret":0}]
```

And I switch the first and the third command, still failed at the third command.

**Update:**
I see miiocli's retry_count=3, timeout=5, so I add `--retries 3` for miot, still failed.

Here's the data captured from wireshark:
```
miiocli:
--> {"id": 2, "method": "get_properties", "params": [{"did": "1044213155", "siid": 2, "piid": 1}]}
<-- {"id":2,"error":{"code":-9999,"message":"user ack timeout"},"exe_time":4004}
--> {"id": 3, "method": "get_properties", "params": [{"did": "1044213155", "siid": 2, "piid": 1}]}
<-- {"id":3,"result":[{"did":"1044213155","siid":2,"piid":1,"value":false,"code":0,"ts":1665232063,"ret":0}],"exe_time":12}

miot:
--> {"method":"set_properties","params":[{"did":"1044213155","value":true,"siid":2,"piid":1}],"id":1}
<-- {"id":1,"error":{"code":-9999,"message":"user ack timeout"},"exe_time":4005}
```

Seems miot retry not correctly working.

**Reply by merdok** [reply]
Oh, that is a very good info! Thanks for that! Will check that out and see if i can improve it!

**Comment by 0x5e** [comment]
A little summary for now:
- Some device it self have trouble when recevice commands constantly, with error: `{"code":-9999,"message":"user ack timeout"}`. To control these device much more reliable, we need to retry after receving error. The retry count can be 2 or 3. Usually the second try works (in my device). #355 
- The root cause I say mesh device have wired behavior before is the property not polling as expected. So the switch status in homekit is always wrong, then I can never send the `Off` command because it is already `Off`, but actually it's `On`.  #357 
- Currently the plugin doesn't support add two sub devices in the same gateway. Maybe using device id as uuid seed instead of IP will be better? But it's difficult to upgrade the plugin.

**Comment by merdok** [comment]
Thanks for the summary and investigation. That is very helpful, i will have a look at the things and try to improve them to make mesh devices work better.
You right, right now the ip and token is used for device identification, which in the case of a mesh device is the gateway hence that will not work, i might add the optional device id for identification which would solve that.

**Reply by 0x5e** [reply]
You are welcome, thanks for your plugin too :)

**Comment by merdok** [comment]
@0x5e In version 1.4.0 i included the deviceId into the accessory id generator which means that you should now be able to add multiple mesh devices to the plugin.

**Reply by 0x5e** [reply]
> @0x5e In version 1.4.0 i included the deviceId into the accessory id generator which means that you should now be able to add multiple mesh devices to the plugin.

Thanks, I will try 😄

**Comment by pedromcortereal** [comment]
hello 0x5e
i'm trying to comunicate with my ble devices (Temperature & humidity sensor (cleargrass.sensor_ht.dk1) and door sensor (isa.magnet.dw2hl)) and i'm not having any success.

my gateway is the Mi Smart Home Hub (lumi.gateway.mgl03)

when i use the command 

miot send $GATEWAY_IP --token $GATEWAY_TOKEN get_properties "[{\"did\":\"$DEVICE_ID\",\"siid\":2,\"piid\":1}]"

i get the following error

INFO Connecting to device at 192.168.1.XXXX
INFO Device found! Sending command: get_properties [{"did":"XXXX","siid":2,"piid":1}]
(Protocol) 192.168.1.XXXX <- Reached maximum number of retries, giving up get_properties - [{"did":"XXXX","siid":2,"piid":1}]
ERROR Call to device timed out

if I try and use the gateway's deviceID i get the gateway's properties. but when i use the ble child devices deviceIDs I always get the error above. it seems that none of the child ble devices of the gateway are reachable through their deviceid...

i noted that the deviceid of those ble child deviced have diferente formats than the "standalone" devices that connect via wifi, like the gateway itself, mi lightbulbs, mi space heater, etc... the latter have deviceids composed by 9 numbers, and the ble devices have deviceids like the following example "blt.2.13b07ds6cog83". is this not the correct deviceID?

**Reply by 0x5e** [reply]
I didn't tried ble device, I guess ble and ble mesh there's different, maybe your experiment shows ble device is not avaliable from gateway :(

**Reply by pedromcortereal** [reply]
i can access the ble device through the gateway as seen below.

![IMG_15EF149BEC40-1](https://user-images.githubusercontent.com/120459606/207825694-a0794656-30a0-477f-89f1-8b368e11b150.jpeg)
